Refactor shiny.fluent to use the updated shiny.react

kamilzyla opened the pull request:
This is a parallel PR to [shiny.react#7](https://github.com/Appsilon/shiny.react/pull/7).

marekrogala commented on inst/examples/Calendar.R:
```diff
@@ -1,23 +1,10 @@
 library(shiny.fluent)
 
 if (interactive()) {
-  dayPickerStrings <- jsonlite::fromJSON(
```
why don't we need this anymore?

kamilzyla replied:
We never needed this - if the `dayPickerStrings` prop is not provided, default settings are used.

marekrogala reviewed (changes requested):
Congrats! :tada: All examples work briliantly.

![50936701_976950922497864_4810986498260205568_n](https://user-images.githubusercontent.com/1421503/116080921-e9580900-a699-11eb-8c83-517bf2c71a59.gif)


 I have a few comments, questions and requests below.


This is a prelimiary review, i didn't look much at the code, just did basic testing.




Adding all comments from examples testing here, but some of them may actually be related more to shiny.react itself.

1. Why rename `srcjs` to `nodejs` ? the former seems like a standard in R packages and apps developed at Appsilon
2. I think this previously was not wrapped in a list. Should it be?
![Selection_625](https://user-images.githubusercontent.com/1421503/116076223-3d5fef00-a694-11eb-8369-04d00298d7bf.png)
3. I have mixed feelings regarding the `.i` suffix -- I really don't like it's ambiguity. Maybe we can have `Checkbox.shinyInput`? more keystrokes, but at least no magic. It will be more clear to people what's the difference between `Checkbox` and `Checkbox.shinyInput` than between `Checkbox` and `Checkbox.i`. Also, this may be especially confusing with buttons -- what should be do about them? `.shinyButton`?
4. In the components documentation dashboard (`examples/dashboard`) looks like the documentation is no longer picked up 
![image](https://user-images.githubusercontent.com/1421503/116078527-04754980-a697-11eb-8ce6-1dc787730751.png)
 5. I think we need examples of the new style of using components (like how do I use `Checkbox` not `Checkbox.i`, `Calendar` not `Calendar.i` etc), right?
 6. I found 1 thing that was not working: `examples/tutorial/dashboard_layout.R` fails due to a missing function `shiny.react::enable_react_debug_mode()`. Looking at master, this function is there -- perhaps you branched off before this change was in shiny.react? or did you remove it?

kamilzyla commented:
Marek, thanks a lot for your effort! :slightly_smiling_face: Answers for your questions:
1. Actually, `srcjs` -> `node/src`. I wanted to put all files related to Node into a single directory to keep things tidy. I figured `node` suits this intention better and I didn't want to have `srcjs/src`.
2. Yes, I intentionally changed the ComboBox interface. This way the text entered is always available as `$text`, and if it's one of the predefined options - you can also access `$key`. I think this interface is easier and more natural to use. Previously you had to use `is.character` to differentiate between custom text and predefined option.
3. Done. I used `.shinyInput` for buttons as well - they are also inputs (they take `inputId` and can be observed via `input$` in the server).
4. Fixed.
5. We'll definitely need additional examples. I'm thinking of a vignette explaining the difference between e.g. plain components and `.shinyInput` components. I'd like to do it in a separate PR though :wink:.
6. Fixed.

kamilzyla commented on inst/examples/dashboard/example_page.R:
```diff
@@ -17,9 +17,14 @@ Rd2list <- function(Rd) {
+  # myrd <- utils:::.getHelpFile(thefile)
+  # Rd2list(myrd)
+  list(
+    description = "TODO",
```
Fixed.

kamilzyla commented on inst/examples/dashboard/www/examples_scripts.js:
```diff
@@ -86,26 +81,29 @@ const keytipConfig = {
 
 // Marquee
+const React = modules['react'];
+const Fluent = modules['@fluentui/react'];
```
Thanks :slightly_smiling_face:

kamilzyla commented on node/webpack.config.babel.js:
```diff
@@ -0,0 +1,30 @@
+    ],
+  },
+  externals: {
+    '@/shiny.react': 'modules["@/shiny.react"]',
```
Instead of variously named global variables for accessing modules (`React` for `react`, `ReactDOM` for `react-dom`, `window['shiny.react']` for our own `shiny.react` code), I want to introduce the following convention:
1. NPM module `name` is available as `jsmodule['name']`,
2. Custom JS bundled with R package `name` is available as `jsmodule['@/name']`.

kamilzyla replied:
I'm going to include this info in a vignette for shiny.react. Note: I changed the name from `modules` to `jsmodule` after discovering that the former would prevent custom tiles from working in Leaflet (due to a bug in Leaflet, actually).

kamilzyla commented on inst/examples/002_more_inputs/app.R:
```diff
@@ -32,84 +31,79 @@ server <- function(input, output, session) {
+    outputName <- paste0("text", i)
     output[[outputName]] <- renderText({
-      sprintf("Value: %s", input[[inputName]])
+      value <- capture.output(dput(input[[inputName]]))
```
It is not needed, but `dput()` provides more useful output for this example. For value being `NULL`, `FALSE`, or `"FALSE"` we get `value: NULL`, `value: FALSE` and `value: "FALSE"` respectively - instead of empty string in the first case and `value: FALSE` in the other two cases.

Unfortunately, `dput()` can only print to file / standard output - that's what the `capture.output` is for. I extracted this into a function.

marekrogala commented on install.sh:
```diff
@@ -2,10 +2,10 @@
 
 # Install the JavaScript dependencies
-yarn
+(cd node && yarn)
```
1. > Actually, srcjs -> node/src. I wanted to put all files related to Node into a single directory to keep things tidy. I figured node suits this intention better and I didn't want to have srcjs/src.

How about `js` and `js/src` ? I don't like `node` because this suggests there's some node application and basically node is not the important point here. This is the JS part of our codebase, not Node part of our codebase -- we're just using node to build the JS part of the frontend


2. since we're making changes to this path and it is repeated, I suggest making this a variable

kamilzyla replied:
Done.

marekrogala commented on node/src/inputs.jsx:
```diff
@@ -0,0 +1,96 @@
+import * as Fluent from '@fluentui/react';
+import { ButtonAdapter, InputAdapter } from '@/shiny.react';
```
what does `from '@/shiny.react'` mean? (specifically `@/`?)

marekrogala commented on node/src/inputs.jsx:
```diff
@@ -0,0 +1,96 @@
+export const IconButton = ButtonAdapter(Fluent.IconButton);
+export const PrimaryButton = ButtonAdapter(Fluent.PrimaryButton);
+
+export const Calendar = InputAdapter(Fluent.Calendar, (value, setValue) => ({
```
so this is created manually, is that correct? but we're still using some of the work that the generator did, like documentation?

kamilzyla replied:
Correct.

marekrogala commented on R/inputs.R:
```diff
@@ -0,0 +1,163 @@
+  )
+}
+
+#' @rdname Button
```
how did you generate this file? anything automated or purely manually?

kamilzyla replied:
Manually. I'd like to decrease the amount of repetitive typing required, but using R's meta-programming capabilities.

marekrogala commented on R/data.R:
```diff
@@ -1,11 +1,11 @@
 #'
 #' @source \url{https://developer.microsoft.com/en-us/fluentui#/controls/web/peoplepicker}
-"fluent_people"
+"fluentPeople"
```
interesting, why this change? in general the snake_case is the standard in R, that's why I made it this way. Unfortunately Shiny chose a different convention and therefore for UI components we used camelCase

marekrogala commented on inst/www/shiny.fluent/shiny-fluent.js.LICENSE.txt:
```diff
@@ -0,0 +1,14 @@
+/*! *****************************************************************************
```
hmm, is this in the right place? I mean, it looks as if this file is a license for the entire shiny.fluent, not only for the Fluent UI. It's named `shiny-fluent.js.LICENSE.txt `... Why do we need this?

kamilzyla replied:
Webpack generates this file along `shiny-fluent.js`, as it has Fluent UI bundled in it.

marekrogala reviewed (approved):
Looks great! Thanks for the detailed split into commits, made the review much smoother.
